File: app/src/main/java/com/polok/eubmanagement/base/BaseFragment.java
package com.polok.eubmanagement.base;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.viewbinding.ViewBinding;
import com.polok.eubmanagement.base.event.Event;
import com.polok.eubmanagement.util.SharedPref;
import com.polok.eubmanagement.widget.PrimaryLoader;

public abstract class BaseFragment<VB extends ViewBinding> extends Fragment {
    protected VB binding;
    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        SharedPref.init(context);
    }
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = initViewBinding();

        initViewModel();

        initOnCreateView(savedInstanceState);

        if (initViewModel() != null) {
            initViewModel().getMessageEvent().observe(getViewLifecycleOwner(), new Event.EventObserver<>(message -> {
                if (message != null && !message.isEmpty()) {
                    Toast.makeText(getContext(), message, Toast.LENGTH_SHORT).show();
                }
            }));
            initViewModel().getLoadingEvent().observe(getViewLifecycleOwner(), new Event.EventObserver<>(isLoading -> {
                if (isLoading != null && initPrimaryLoader() != null) {
                    if (isLoading) initPrimaryLoader().setVisibility(View.VISIBLE);
                    else initPrimaryLoader().setVisibility(View.GONE);
                }
            }));
            initViewModel().getNavigateEvent().observe(getViewLifecycleOwner(), new Event.EventObserver<>(onNavigate -> {
                if (onNavigate != null) onNavigateEvent(onNavigate.getId(), onNavigate.getBundle());
            }));
        }

        return binding.getRoot();
    }

    protected abstract VB initViewBinding();

    protected abstract BaseViewModel initViewModel();

    protected abstract void initOnCreateView(Bundle savedInstanceState);

    protected abstract PrimaryLoader initPrimaryLoader();

    protected void onNavigateEvent(int id, Bundle bundle) {}

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/polok/eubmanagement/base/BaseFragment.kt
package com.polok.eubmanagement.base

import android.annotation.SuppressLint
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.viewbinding.ViewBinding
import com.polok.eubmanagement.base.event.EventObserver
import com.polok.eubmanagement.util.SharedPref
import com.polok.eubmanagement.util.hideSoftKeyBoard
import com.polok.eubmanagement.util.makeGone
import com.polok.eubmanagement.util.makeVisible
import com.polok.eubmanagement.util.showToast
import com.polok.eubmanagement.widget.PrimaryLoader

abstract class BaseFragment<VB: ViewBinding>(
    viewBindingFactory: (LayoutInflater) -> VB
) : Fragment() {

    protected val binding: VB by lazy {
        viewBindingFactory(layoutInflater)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        SharedPref.init(context)
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        return binding.root.apply {
            setOnTouchListener { view, _ ->
                view.hideSoftKeyBoard()
                return@setOnTouchListener false
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initOnCreateView(savedInstanceState)

        initViewModel()?.messageEvent?.observe(viewLifecycleOwner, EventObserver {
            if (it.isNotEmpty()) context.showToast(it)
        })
        initViewModel()?.loadingEvent?.observe(viewLifecycleOwner, EventObserver {
            if (initPrimaryLoader() != null) {
                if (it) initPrimaryLoader().makeVisible() else initPrimaryLoader().makeGone()
            }
        })
        initViewModel()?.navigateEvent?.observe(viewLifecycleOwner, EventObserver {
            onNavigateEvent(it.id, it.bundle)
        })
    }

    protected abstract fun initViewModel(): BaseViewModel?

    protected abstract fun initOnCreateView(savedInstanceState: Bundle?)

    protected abstract fun initPrimaryLoader(): PrimaryLoader?

    protected open fun onNavigateEvent(id: Int, bundle: Bundle?) {}
}